src/main.rs

```diff
@@ -285,77 +286,108 @@ fn main() {
+                message = "Reloading config.",
+                path = ?opts.config_path
+            );
 
-        let file = if let Some(file) = open_config(&opts.config_path) {
-            file
-        } else {
-            continue;
+            let file = if let Some(file) = open_config(&opts.config_path) {
+                file
+            } else {
+                continue;
+            };
+
+            trace!("Parsing config");
+            let config = vector::topology::Config::load(file);
+            let config = handle_config_errors(config);
+            if let Some(config) = config {
+                let success =
+                    topology.reload_config_and_respawn(config, &mut rt, opts.require_healthy);
+                if !success {
+                    error!("Reload was not successful.");
+                }
+            } else {
+                error!("Reload aborted.");
+            }
         };
 
-        trace!("Parsing config");
-        let config = vector::topology::Config::load(file);
-        let config = handle_config_errors(config);
-        if let Some(config) = config {
-            let success = topology.reload_config_and_respawn(config, &mut rt, opts.require_healthy);
-            if !success {
-                error!("Reload was not successful.");
+        if signal == SIGINT || signal == SIGTERM {
+            use futures::future::Either;
+
+            info!("Shutting down.");
+            let shutdown = topology.stop();
+            metrics_trigger.cancel();
+
+            match rt.block_on(shutdown.select2(signals.into_future())) {
+                Ok(Either::A(_)) => { /* Graceful shutdown finished */ }
+                Ok(Either::B(_)) => {
+                    info!("Shutting down immediately.");
+                    // Dropping the shutdown future will immediately shut the server down
+                }
+                Err(_) => unreachable!(),
             }
+        } else if signal == SIGQUIT {
+            info!("Shutting down immediately");
+            drop(topology);
         } else {
-            error!("Reload aborted.");
+            unreachable!();
         }
-    };
+    }
+    #[cfg(windows)]
+    {
+        let mut ctrl_c_stream = tokio_signal::ctrl_c().flatten_stream();
+        let ctrl_c = future::poll_fn(move || ctrl_c_stream.poll());
+        let crash = future::poll_fn(move || graceful_crash.poll());
+
+        let interruptions = ctrl_c.select2(crash);
+        rt.block_on(interruptions)
+            .map_err(|_| ())
+            .expect("Neither stream errors");
 
-    if signal == SIGINT || signal == SIGTERM {
         use futures::future::Either;
 
         info!("Shutting down.");
         let shutdown = topology.stop();
         metrics_trigger.cancel();
 
-        match rt.block_on(shutdown.select2(signals.into_future())) {
+        let ctrl_c_stream = tokio_signal::ctrl_c().flatten_stream();
+        match rt.block_on(shutdown.select2(ctrl_c_stream.into_future())) {
```

Comment:
can we just use `block_on` for both windows and unix instead of having two code paths here?

Reply:
We still need two code paths because Unix code also contains signals handling.

Reply:
I'd like the `main.rs` to contain as little platform specific things as possible and be clear about just a shutdown signal. So if we can combine as much code as possible and reduce the `cfg` that would be good but we can also move some of this code to `signal.rs` file. Just seems like a lot of code here for the main file.